### PR TITLE
External int

### PR DESCRIPTION
**Pull request by conorpp**
Adds `InputMux` and `Pint` peripheral wrappers.  Example added for setting up external interrupt on a pin.

### CODE REVIEW

**Comment by conorpp** [comment]
Depends on https://github.com/nickray/lpc55-pac/pull/4

File: examples/external_interrupts.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#![no_main]
+#![no_std]
+
+extern crate panic_semihosting;  // 4004 bytes
```
Comment (nickray) [inline examples/external_interrupts.rs]: That's "old" syntax, these days (Rust edition 2018: https://doc.rust-lang.org/edition-guide/rust-2018/index.html) just do `use panic_semihosting;`. Just FYI, I see some of my examples have that too ;)

File: src/lib.rs [hunk 2]
```diff
@@ -147,18 +149,24 @@ pub struct Peripherals {
     /// Micro-Tick Timer
     pub utick: Utick,
 
-    /// Analog-to-Digital Converter (ADC) - not HAL-ified.
-    pub ADC0: raw::ADC0,
+    /// Input multiplexer
+    pub inputmux: InputMux,
```
Comment (nickray) [inline src/lib.rs]: Can we keep these sorted alphabetically? First "wrapped" peripherals alphabetically, then "raw"/passthrough peripherals alphabetically.

**Comment by conorpp** [comment]
Should be ready to merge

File: examples/external_interrupts.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    let mut pint = hal.pint.enabled(&mut hal.syscon);
+
+    pint.enable_interrupt(&mut mux, &input, 0, Mode::RisingEdge);
+    pint.enable_interrupt(&mut mux, &input, 0, Mode::FallingEdge);
+
+    // // Dont need mux anymore
+    mux.disabled(&mut hal.syscon);
+
+    // Clear interrupts initially
+    pint.rise.write(|w| unsafe { w.bits(1) });
```
Comment (nickray) [inline examples/external_interrupts.rs]: Gosh I never realized you could use the Deref outside the implementation too!

File: src/peripherals/pint.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    /// LPC55 supports 8 external pin interrupts, from any PIO pin.
+    /// Use `slot` to indicate (0-7) which slot you'd like to use.
+    /// `mode` indicates what kind of interrupt to generate.
+    /// You can call this function twice to enable both `RisingEdge` and `FallingEdge` interrupts for same pin + slot.
+    pub fn enable_interrupt<PIN: PinId>(
+        &mut self, 
+        mux: &mut InputMux<init_state::Enabled>, 
+        _pin:  &Pin<PIN, state::Gpio<direction::Input>>, 
+        slot: u8, mode: Mode
```
Comment (nickray) [inline src/peripherals/pint.rs]: Maybe it makes sense to have a `Slot` enum?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=894492b474f3cbf9cc0d328a44544ebb

Or at least a check that slot is <= 8.

Comment (nickray) [reply src/peripherals/pint.rs]: Also I think UM calls it slice, not slot?

Comment (conorpp) [reply src/peripherals/pint.rs]: Done!  slice is more for pattern matching, and a "slot"/pin0-7 still needs to be selected as input for a slice.

**Comment by conorpp** [comment]
New comment to use Slot enum has been addressed

**Comment by nickray** [comment]
Cool, thanks, and sorry for the delay in merging.

**Pull request merged**